### DISCUSSION TITLE
Decimal Registers

### DISCUSSION THREAD

**Discussion by adesutherland** [adesutherland/CREXX#364] [General]
The following are extracted from the code and are all the float rxas instructions. For each of these we need the decimal equivalent (with a d not a f). 

This way we keep consistency - and minimise compiler changes. We also need a OP_DEC define (for literals - later)

And we need a setdigits OP_REG and setdigits OP_INT to set the number of significant digits the engine should use from now on (until the next call to setdigits)

Also I think we should haven in the register definition, a significant digits: 

     decNumber *decimal_value;
     int sig_digits;

(unless the decNumber structure already has this information?) so that the decimal maths function can make sure that both operators have the correct - well - memory size, I guess. For example, the system might want to calculate a sum to 20 sig_digits - it will have to ensure the buffer of the result is big enough but also increase the buffer size of the two operands (maybe?)

This list also needs the **dadd** instruction for addition. I forgot about this when I made the list originally

    instr_f("fsub", "Float Subtract (op1=op2-op3)", OP_REG, OP_REG, OP_REG);
    instr_f("fsub", "Float Subtract (op1=op2-op3)", OP_REG, OP_REG, OP_FLOAT);
    instr_f("fsub", "Float Subtract (op1=op2-op3)", OP_REG, OP_FLOAT, OP_REG);

    instr_f("fmult", "Float Multiply (op1=op2*op3)", OP_REG, OP_REG, OP_REG);
    instr_f("fmult", "Float Multiply (op1=op2*op3)", OP_REG, OP_REG, OP_FLOAT);

    instr_f("fdiv", "Float Divide (op1=op2/op3)", OP_REG, OP_REG, OP_REG);
    instr_f("fdiv", "Float Divide (op1=op2/op3)", OP_REG, OP_REG, OP_FLOAT);
    instr_f("fdiv", "Float Divide (op1=op2/op3)", OP_REG, OP_FLOAT, OP_REG);

    instr_f("feq", "Float Equals op1=(op2==op3)", OP_REG, OP_REG, OP_REG);
    instr_f("feq", "Float Equals op1=(op2==op3)", OP_REG, OP_REG, OP_FLOAT);
    instr_f("fne", "Float Not equals op1=(op2!=op3)", OP_REG, OP_REG, OP_REG);
    instr_f("fne", "Float Not equals op1=(op2!=op3)", OP_REG, OP_REG, OP_FLOAT);
    instr_f("fgt", "Float Greater than op1=(op2>op3)", OP_REG, OP_REG, OP_REG);
    instr_f("fgt", "Float Greater than op1=(op2>op3)", OP_REG, OP_REG, OP_FLOAT);
    instr_f("fgt", "Float Greater than op1=(op2>op3)", OP_REG, OP_FLOAT, OP_REG);
    instr_f("fgte", "Float Greater than equals op1=(op2>=op3)", OP_REG, OP_REG, OP_REG);
    instr_f("fgte", "Float Greater than equals op1=(op2>=op3)", OP_REG, OP_REG, OP_FLOAT);
    instr_f("fgte", "Float Greater than equals op1=(op2>=op3)", OP_REG, OP_FLOAT, OP_REG);
    instr_f("flt", "Float Less than op1=(op2<op3)", OP_REG, OP_REG, OP_REG);
    instr_f("flt", "Float Less than op1=(op2<op3)", OP_REG, OP_REG, OP_FLOAT);
    instr_f("flt", "Float Less than op1=(op2<op3)", OP_REG, OP_FLOAT, OP_REG);
    instr_f("flte", "Float Less than equals op1=(op2<=op3)", OP_REG, OP_REG, OP_REG);
    instr_f("flte", "Float Less than equals op1=(op2<=op3)", OP_REG, OP_REG, OP_FLOAT);
    instr_f("flte", "Float Less than equals op1=(op2<=op3)", OP_REG, OP_FLOAT, OP_REG);
    instr_f("fgtbr", "Float Greater than if (op2>op3) goto op1", OP_ID, OP_REG, OP_REG);
    instr_f("fltbr", "Float Less than if (op2<op3) goto op1", OP_ID, OP_REG, OP_REG);

    instr_f("ret", "Return op1", OP_FLOAT, OP_NONE, OP_NONE);

    instr_f("fcopy", "Copy Float op2 to op1", OP_REG, OP_REG, OP_NONE);

    instr_f("load", "Load op1 with op2", OP_REG, OP_FLOAT, OP_NONE);

    instr_f("say", "Say op1", OP_FLOAT, OP_NONE, OP_NONE);

    instr_f("btof", "Set register float value from its boolean value", OP_REG, OP_NONE, OP_NONE);
    instr_f("ftos", "Set register string value from its float value", OP_REG, OP_NONE, OP_NONE);
    instr_f("itof", "Set register float value from its int value", OP_REG, OP_NONE, OP_NONE);
    instr_f("ftoi", "Set register int value from its float value", OP_REG, OP_NONE, OP_NONE);
    instr_f("ftob", "Set register boolean (int 1 or 0) value from its float value", OP_REG, OP_NONE, OP_NONE);
    instr_f("stof", "Set register float value from its string value", OP_REG, OP_NONE, OP_NONE);
   
    instr_f("fformat", "Set string value from float value using a format string", OP_REG, OP_REG, OP_REG);

    instr_f("fpow", "op1=op2**op3", OP_REG, OP_REG, OP_REG);
    instr_f("fpow", "op1=op2**op3", OP_REG, OP_REG, OP_FLOAT);

    instr_f("fsex", "float op1 = -op1 (sign change)", OP_REG, OP_NONE,OP_NONE);

**Comment by Peter-Jacob** [comment]
Thanks, Adrian, but this means we go beyond the simple register instructions for DEC that we discussed yesterday.
example: `instr_f("ddiv", "Decimal Divide (op1=op2/op3)", OP_REG, OP_REG, OP_FLOAT);`
instead of `instr_f("ddiv", "Decimal Divide (op1=op2/op3)", OP_REG, OP_REG, OP_REG);` where I expect the float is already migrated to DEC
This is just to make sure there are no misunderstandings.

**Comment by Peter-Jacob** [comment]
` We also need a OP_DEC define (for literals - later)`
What should it do?

**Comment by adesutherland** [comment]
Hi Peter - answering your questions together.

Now: We just need - instr_f("ddiv", "Decimal Divide (op1=op2/op3)", OP_REG, OP_REG, OP_REG) where all the registers have dec values.

Future: we **MAY** have instr_f("ddiv", "Decimal Divide (op1=op2/op3)", OP_REG, OP_REG, **OP_DEC**), in which case we will need to define what a literal decimal is (OP_DEC), what it looks like in rxas and how it is coded in the binary and in the interpreter program itself. Not sure - and not needed yet

**Reply by adesutherland** [reply]
I am thinking that the decimal literals, e.g. in rxas

    dadd r3,r2,1.3    # Note just using the float literal syntax here

would be stored as decPacked format in the constant pool. So the instruction would have a pointer to the decPacked and convert to  decNumber.

In sum, be ready for this - the instructions with OP_DEC will get a pointer to a decPacked that is converted back into a decNumber
